### PR TITLE
Create IMPLEMENTATION_SPEC.md

### PR DESCRIPTION
**Pull request by musnit**

### CODE REVIEW

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+The spec is written under the assumption that the LICENCE for Peggy will be Apache 2.0, like the Cosmos SDK, or that all code can be easily used. The project has not been completed nor updated in the last few months. I suggest contacting the Peggy team and possibly even considering asking for a grant to do the work of getting it completed.
+
+## Relation to main Bifrost paper
+The main Bifrost paper describes a more complex and possibly more secure system with a custom `m of n` validator signatures required, support for a changing validator set, random reshuffling and VRFs, SPV proofs and gossip-layer communication. The spec is constrained to a very simple MVP that can be used for a Testnet and get real users running and playing with it asap, and does not include any of those features or security/performance improvements.
+
+## Overview
+The Bifrost MVP will add functionality to THORChain that allows the movement of Ether and ERC20 from Ethereum to THORChain and back, where they are represented as ERC20 tokens. It will not support movement of RUNE or THORChain native tokens to Ethereum.
+
+The MVP is split into three components:
```
Comment (jkbrooks) [inline x/eth-bifrost/IMPLEMENTATION_SPEC.md]: four components, not three

**Comment by jkbrooks** [comment]
I don't have any major feedback as I think this is good for now.  I also think we will need to specify UI elements, specific parameters, and a plan for switching from Infura to our own Ethereum nodes but this is not essential for an MVP

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+ - the Ethereum smart contract for sending tokens into the Bifrost from Ethereum
+ - the current THORChain Full Node and Validator, with additional bifrost and oracle modules
+ - a signer process, that detects Bifrost requests to send to Ethereum, and signs them appropriately for Ethereum
```
Comment (philipstanislaus) [inline x/eth-bifrost/IMPLEMENTATION_SPEC.md]: Why do we need the signer process? A tx that is in the chain has already be signed by >2/3 of our validators. Isn't the tx with the signatures (maybe a smaller merkle-proof package) enough proof for the smart contract to unlock/mint?

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ - Quick audit of cosmos Oracle module to determine what can be used and what needs to be fixed
+ - Completion of Oracle module functionality
+ - Development of Bifrost module that accepts actions from the oracle module and mints tEth/tERC20 on THORChain and controls whitelisting
+ - Development of relayer service that watches Ethereum via Infura and submits transactions to the Oracle module
+ - Tests and QA of transfers from Ether/ERC20 into THORChain
+ - Enhance Bifrost module to be able to accept and burn tokens from users for sending back over the Bifrost to Ethereum
+ - Development of signer service that watches Bifrost module for these transactions, generates Ethereum signatures and submits them back to the Bifrost module for future relaying (or, to the reverse-oracle module if that path is explored)
+ - Enhance relayer to detect `m of n` signatures (or, to detect a `m of n` reverse-oracle state change if that path is explored) and relay them to the Ethereum smart contract
+ - Tests and QA of transfers of tEth/tERC20 on THORChain back to Ethereum
+ - Special rules for the official RUNE ERC20 contract to work differently to generic ERC20 contracts implemented
```
Comment (philipstanislaus) [inline x/eth-bifrost/IMPLEMENTATION_SPEC.md]: Which rules do we need here?

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+The spec is written under the assumption that the LICENCE for Peggy will be Apache 2.0, like the Cosmos SDK, or that all code can be easily used. The project has not been completed nor updated in the last few months. I suggest contacting the Peggy team and possibly even considering asking for a grant to do the work of getting it completed.
+
+## Relation to main Bifrost paper
+The main Bifrost paper describes a more complex and possibly more secure system with a custom `m of n` validator signatures required, support for a changing validator set, random reshuffling and VRFs, SPV proofs and gossip-layer communication. The spec is constrained to a very simple MVP that can be used for a Testnet and get real users running and playing with it asap, and does not include any of those features or security/performance improvements.
```
Comment (philipstanislaus) [inline x/eth-bifrost/IMPLEMENTATION_SPEC.md]: Support for changing validators is in Peggy too.

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+## Overview
+The Bifrost MVP will add functionality to THORChain that allows the movement of Ether and ERC20 from Ethereum to THORChain and back, where they are represented as ERC20 tokens. It will not support movement of RUNE or THORChain native tokens to Ethereum.
```
[the remark this answers is not in the document]
Comment (musnit) [reply x/eth-bifrost/IMPLEMENTATION_SPEC.md]: mostly just the solidit of peggy looks testnet-ready, the rest of the code not so much. so for mvp i don't think worrying about that is necessary. also i think for RUNE there was mention of us wanting to intentionally not support that, but to confirm with JP.

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+The MVP is split into three components:
+ - the Ethereum smart contract for sending tokens into the Bifrost from Ethereum
+ - the current THORChain Full Node and Validator, with additional bifrost and oracle modules
+ - a signer process, that detects Bifrost requests to send to Ethereum, and signs them appropriately for Ethereum
+ - a relayer process, that detects transactions to relay between THORChain and Ethereum
```
[the remark this answers is not in the document]
Comment (musnit) [reply x/eth-bifrost/IMPLEMENTATION_SPEC.md]: yeh peggy's witness stuff had a few concerns
 - i got the impression it was written mostly before they seperated the oracle idea into its own module, and so was mixing ideas with that a bit
 - parts of it are related to communication between normal cosmos zone and tendermint zone, whereas in our scenario we are just doing both the peg functionality and everything else in a single zone on thorchain, so simpler without any ibc
 - for simplicity just called both directions a relayer, even though logic/code for each direction may be a bit different. may be worth considering still splitting this into two like a witness vs relayer but yeh i'll leave that up to implementors for what they prefer, realize when implementing

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+Alice started with 10 Ether on Ethereum. She sent all of them to THORChain. There she sent 4 tEth to Bob. Bob then redeemed those 4 tEth back to Ethereum.
+
+The end result is that Alice holds 6 tEth and 0 Ether and Bob holds 4 Ether and 0 tEth.
+
+# Components
+
+## Ethereum smart contract component
+
+The smart contracts verify updates coming from THORChain
+using the known keys of the signing apps. The smart contracts
```
[the remark this answers is not in the document]
Comment (musnit) [reply x/eth-bifrost/IMPLEMENTATION_SPEC.md]: umm i think can register/unregister them on the smart contract through validator changes, not sure, didn't look too thoroughly through the full solidity code, but i think yes for mvp/v1 we can hardcode

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply x/eth-bifrost/IMPLEMENTATION_SPEC.md]: I talked to @jpthor about that today – he likes the idea of RUNE being as fungible as other tokens, so no exception for it for now.

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md (hunk 2 above)
Comment (philipstanislaus) [reply x/eth-bifrost/IMPLEMENTATION_SPEC.md]: Got it, thanks!

File: x/eth-bifrost/IMPLEMENTATION_SPEC.md (hunk 3 above)
Comment (philipstanislaus) [reply x/eth-bifrost/IMPLEMENTATION_SPEC.md]: Good point, all tokens created on THORChain should probably locked rather than burned, right?

**Pull request merged**